File: boj/DynamicProgramming/BOJ_1932.py
# 백준 1932 정수 삼각형
# 21.02.22

import sys
from functools import cmp_to_key
from collections import deque
from itertools import combinations
from itertools import permutations
sys.setrecursionlimit(10**6)

n = int(input())
d = [[int(sys.stdin.readline())]]
for i in range(1, n):
    d.append(list(map(int, sys.stdin.readline().split())))
    for j in range(len(d[i])):
        if j == 0: d[i][j] += d[i-1][0] # 맨 앞 원소는 바로 위 원소랑 더하기
        elif j == len(d[i]) - 1: d[i][j] += d[i-1][j -1] # 맨 뒤 원소는 바로 위 원소랑 더하기
        else:
            d[i][j] += max(d[i-1][j-1], d[i-1][j])

print(max(d[n-1]))

File: boj/DynamicProgramming/BOJ_1932.swift
// 백준 1932 정수 삼각형
// 21.02.24

import Foundation

let n = Int(readLine()!)!
var tri: [[Int]] = []
tri.append(readLine()!.split(separator: " ").map { Int(String($0))! })

for i in stride(from: 1, to: n, by: 1) {
  tri.append(readLine()!.split(separator: " ").map { Int(String($0))! })
  
  for j in stride(from: 0, through: i, by: 1) {
    if j == 0 { tri[i][j] += tri[i-1][0] }
    else if j == i { tri[i][j] += tri[i-1][j-1] }
    else { tri[i][j] += max(tri[i-1][j-1], tri[i-1][j]) }
  }
}

print(tri[n-1].max()!)
